core/ibft.go

```
@@ -633,84 +634,110 @@ func (i *IBFT) validateProposal0(msg *proto.Message, view *proto.View) bool {
 
 // validateProposal validates a proposal for round > 0
 func (i *IBFT) validateProposal(msg *proto.Message, view *proto.View) bool {
-	var (
-		height = view.Height
-		round  = view.Round
-
-		proposalHash = messages.ExtractProposalHash(msg)
-		certificate  = messages.ExtractRoundChangeCertificate(msg)
-		rcc          = messages.ExtractRoundChangeCertificate(msg)
-	)
+	// Make sure the current node is not the proposer for this round
+	if i.backend.IsProposer(i.backend.ID(), view.Height, view.Round) {
+		return false
+	}
 
 	// Make sure common proposal validations pass
 	if !i.validateProposalCommon(msg, view) {
 		return false
 	}
 
+	certificate := messages.ExtractRoundChangeCertificate(msg)
+
+	// Make sure the certificate is valid
+	if err := i.validateRoundChangeCertificate(certificate, view); err != nil {
+		return false
+	}
+
+	roundsAndPreparedProposalHashes := i.extractRoundsAndProposalHashes(certificate, view)
+	if len(roundsAndPreparedProposalHashes) == 0 {
+		return true
+	}
+
+	// Find the max round and its prepared proposal hash
+	var (
+		maxRound             uint64
+		maxRoundProposalHash []byte
+	)
+
+	for round, hash := range roundsAndPreparedProposalHashes {
+		if round >= maxRound {
+			maxRound = round
+			maxRoundProposalHash = hash
+		}
+	}
+
+	return bytes.Equal(maxRoundProposalHash, messages.ExtractProposalHash(msg))
+}
+
+func (i *IBFT) validateRoundChangeCertificate(
```

Why return an error from this method, if you're only using it as  a boolean value anyways?
Not to mention that the error is not checked anywhere (typed) specifically

---

Added a log line to display the error in question when validating an RCC

[1e0b261](https://github.com/madz-lab/go-ibft/pull/10/commits/1e0b261d86895e437a1518b06634d0e13b6d0a3b)